Add Service Worker static routing API (#34555)
mdn/content: files/en-us/web/api/service_worker_api/index.md

{{DefaultAPISidebar("Service Workers API")}}{{AvailableInWorkers}}

Service workers essentially act as proxy servers that sit between web applications, the browser, and the network (when available). They are intended, among other things, to enable the creation of effective offline experiences, intercept network requests, and take appropriate action based on whether the network is available, and update assets residing on the server. They will also allow access to push notifications and background sync APIs.

## Service worker concepts and usage

A service worker is an event-driven [worker](/en-US/docs/Web/API/Worker) registered against an origin and a path. It takes the form of a JavaScript file that can control the web-page/site that it is associated with, intercepting and modifying navigation and resource requests, and caching resources in a very granular fashion to give you complete control over how your app behaves in certain situations (the most obvious one being when the network is not available).

> A service worker is run in a worker context: it therefore has no DOM access and runs on a different thread to the main JavaScript that powers your app, so it is non-blocking. It is designed to be fully async; as a consequence, APIs such as synchronous [XHR](/en-US/docs/Web/API/XMLHttpRequest) and [Web Storage](/en-US/docs/Web/API/Web_Storage_API) can't be used inside a service worker.

Reviewer: Felt a bit clunky  before and after comma removal. How about.

```suggestion
A service worker is run in a worker context: it therefore has no DOM access and runs on a different thread to the main JavaScript that powers your app. It is non-blocking and are designed to be fully asynchronous. As a consequence, APIs such as synchronous [XHR](/en-US/docs/Web/API/XMLHttpRequest) and [Web Storage](/en-US/docs/Web/API/Web_Storage_API) can't be used inside a service worker.
```

Author: I've basically used this, but I've made some adjustments to sort out the plurality confusion.

If this is the first time a service worker has been made available, installation is attempted, then after a successful installation, it is activated.

> If there is an existing service worker available, the new version is installed in the background, but not yet activated — at this point, it is called the _worker in waiting_. It is only activated when there are no longer any pages loaded that are still using the old service worker. As soon as there are no more pages to be loaded, the new service worker activates (becoming the _active worker_). Activation can happen sooner using {{DOMxRef("ServiceWorkerGlobalScope.skipWaiting()")}} and existing pages can be claimed by the active worker using {{DOMxRef("Clients.claim()")}}.

Reviewer: Yoda, you are not.
```suggestion
If there is an existing service worker available, the new version is installed in the background, but not yet activated — at this point it is called the _worker in waiting_. It is only activated when there are no longer any pages loaded that are still using the old service worker. As soon as there are no more pages to be loaded, the new service worker activates (becoming the _active worker_). Activation can happen sooner using {{DOMxRef("ServiceWorkerGlobalScope.skipWaiting()")}} and existing pages can be claimed by the active worker using {{DOMxRef("Clients.claim()")}}.
```

For a complete tutorial to show how to build up your first basic example, read [Using Service Workers](/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers).

> ### Adding static routing to control how resources are fetched

Reviewer: ```suggestion
### Using static routing to control how resources are fetched
```

> Service workers can incur a performance cost — when a page is loaded for the first time in a while, the browser has to wait for the service worker to start up and run to know what content to load. This cost can be significant, for example, if you are attempting to improve performance via service worker caching strategies.
> 
> In such a case, it would be preferable to bypass the service worker altogether and fetch resources straight from a browser {{domxref("Cache")}}. The {{domxref("InstallEvent.addRoutes()")}} method can be used to implement this use case and more.

Reviewer: This text implies the cost is something to do with your caching strategy, but it is equally true if you are planning on fetching from the network. The cost is in the waiting. So perhaps:

```suggestion
Service workers can incur an unnecessary performance cost — when a page is loaded for the first time in a while, the browser has to wait for the service worker to start up and run to know what content to load and whether it is to come from a cache or the network.

If you can already know ahead of time where content will be particular content will be fetched from, you can bypass the service worker altogether and fetch resources immediately. The {{domxref("InstallEvent.addRoutes()")}} method can be used to implement this use case and more.
```

Author: Used, with a few tweaks.
